soroban-test-wasms/wasm-workspace/Cargo.toml

```diff
@@ -0,0 +1,64 @@
+# This file defines a workspace that is _periodically_ recompiled as part of
+# updating the WASM files that are then embedded as as binary constants in the
+# surrounding crate `soroban-test-wasms`.
+#
+# Neither the rust crates in this workspace, nor the workspace itself, are (in
+# cargo's mind) part of the outer workspace containing
+# soroban-env-{host,guest,macros,wasm-tests} etc. The inner workspace is
+# excluded from the outer workspace, and the crates within it must be manually
+# built (and should only be built as WASM). See the adjacent `Makefile` for
+# invocation.
+#
+# The dependency graph here is complex and subtle and needs to be understood if
+# you're going to use it correctly. There is further explanation in
+# https://rs-soroban-env/soroban-test-wasms/README.md
+
+[workspace]
+resolver = "2"
+
+members = ["add_i32", "create_contract", "invoke_contract", "linear_memory", "vec"]
+
+[patch.crates-io]
+soroban-env-common = { path = "../../soroban-env-common" }
+soroban-env-guest = { path = "../../soroban-env-guest" }
+soroban-env-host = { path = "../../soroban-env-host" }
+soroban-env-macros = { path = "../../soroban-env-macros" }
+soroban-native-sdk-macros = { path = "../../soroban-native-sdk-macros" }
+
+# You should update this rev to the XDR rev used by the outer workspace
+# anytime you're going to recompile this workspace.
+stellar-xdr = { git = "https://github.com/stellar/rs-stellar-xdr", rev = "f4fe3091" }
+
+# You should actually be able to set this to be any existing version of wasmi because we are
+# not building the SDK in host mode, so the "vm" feature is off.
+wasmi = { package = "soroban-wasmi", git = "https://github.com/stellar/wasmi", rev = "7b1f2355" }
+
+# The following SDK patch lines must denote an SDK version that is _compatible_
+# with the adjacent version of soroban-env-guest, selected above. If you are
+# making a change to soroban-env-common and/or soroban-env-guest, and are lucky,
+# you can modify soroban-env-common and/or soroban-env-guest _without_ having to
+# bump this rev, and can just recompile the WASMs here and embed them in your
+# change to the env repo.
```

Comment:
A few clarification / questions: 
1. Any change to the `soroban-env-host` alone does not require recompilation of the wasm files, only changes touching the `-common` and `-guest` does?  Thus "change to the env repo" here does not include change to `soroban-env-host`?
2. How does this relate to the `interface_version` in `meta.rs`? The `interface_version` specifies version of `soroban_env_common::env`, thus is a narrower restriction? I.e. Change to the `Env` requires bumping the `interface_version` which requires recompiling wasm files, but changes in the  `-common` and `-guest` without touching the `Env` may not need to recompile? 
3. The `sdk` also specifies the commit hash of the env repo: https://github.com/stellar/rs-soroban-sdk/blob/main/Cargo.toml#L45-L49. When recompiling with local sdk, do I need to change these lines to point to the local env repo?

Reply:
1. Correct, these tests are built into WASM, which does not see the host crate at all; it only knows about the guest crate.
2. The interface version number should be changed anytime _anything_ that invalidates the interface between guest and host happens; changes to the Env trait are a minimal example of that, but (as a different example) suppose the bit-packing of RawVal changed, or the enum values of SCVal -- these would represent a "break to the interface" just as surely, and would therefore need to bump the interface version number. We need to be conservative about the interface version number, and bump it _whenever_ we have reason to believe there might be breakage. In terms of how this affects recompiling tests: a bump to the interface version necessarily means you have to recompile the tests; they won't run if the interface version compiled-in to them differs from the version supported on the host. The bigger question is whether the tests actually still compile at all, or whether they need to be updated (and/or the SDK needs to be updated). There is a sequence of things to ensure here, but the tests should either fail to compile, or fail to run, or both if there is a step missed.
3. No, the whole point of this change is that you do not need to use a local SDK to recompile at all. We're referring to a _remote_ SDK (by published-on-github git hash) and hoping it's still compatible with the local, updated env-common / env-guest (which we're injecting into its dependency tree, overriding the SDK's own `rev=` spec for the env repo).

Reply:
(A thing to understand here is that this comment and the crate references the file is defining are within a `[patch.crates-io]` section of this workspace: a section that declares _overrides_ of all the transitive crate references of everything in the workspace; this is a feature that exists specifically for this sort of "build X but with my own override for its dependency on Y" scenario)